src/scicloj/02_numpy.clj

```diff
@@ -216,6 +269,141 @@ Ex.: `x[4]`, `x[-1]`, `x[1,2]`"]
 
 ;; see https://cnuernber.github.io/dtype-next/tech.v3.datatype.html#var-sub-buffer - not sure if possible to somehow define the "step" - perhaps using argops and indexed-buffer
 
+(def x (dtt/->tensor (range 20)))
+
+(defn ravel [t]
+  (let [dt    (dtype/get-datatype t)
+        shape (dtype/shape t)
+        idx'  (index shape)]
+    (-> (dtype/make-reader dt (apply * shape) (apply dtt/mget t (nth idx' idx)))
+        (dtt/->tensor))))
+
+(dtype/shape x)
+(index (dtype/shape x))
+(ravel x)
+
+(defn mask [f t]
+  (-> (dtype/emap #(if (f %) 1 0) :int32 t)
+      dtt/->tensor
+      ravel))
+
+(mask #(= 1 %) x)
+
+
+(defn compress [lhs rhs]
+  (let [lhr   (ravel lhs)
+        rhr   (ravel rhs)
+        dt    (dtype/get-datatype rhs)
+        shape (dtype/shape lhs)
+        idx'  (into []
+                    (comp
+                     (partition-all 2)
+                     (map #(repeat (first %) (second %)))
+                     cat)
+                    (interleave (ravel lhs) (index shape)))
+        size  (count idx')]
+    (-> (dtype/make-reader dt size (apply dtt/mget rhr (nth idx' idx)))
+        (dtt/->tensor))))
+
+
+(compress (mask (partial = 1) x) x)
+
+
+(defn ^:shameful start-stop-step-reverse [t start stop step]
+  (cond
+    (and start stop step (pos? step)) ;; x[1:10:2]
+    [start stop step false]
+    (and start stop (nil? step)) ;; x[1:10]
+    [start stop 1 false]
+    (and start stop step (neg? step)) ;; x[1:10:-1]
+    [stop start (Math/abs step) true]
+    (and start (nil? stop) step (pos? step)) ;; x[1::2]
+    [start (first (dtype/shape t)) step false]
+    (and start (nil? stop) (nil? step)) ;; x[1::]
+    [start (first (dtype/shape t)) 1 false]
+    (and start (nil? stop) step (neg? step)) ;; x[1::-2]
+    [0 start (Math/abs step) true]
+    (and (nil? start) stop step (pos? step)) ;; x[:5:2]
+    [0 stop step false]
+    (and (nil? start) stop (nil? step)) ;; x[:5]
+    [0 stop 1 false]
+    (and (nil? start) stop step (neg? step)) ;; x[:5:-2]
+    [stop (dec (first (dtype/shape t))) (Math/abs step) true]
+    (and (nil? start) (nil? stop) step (pos? step)) ;; x[::2]
+    [0 (first (dtype/shape t)) step false]
+    (and (nil? start) (nil? stop) (nil? step)) ;; x[::]
+    [0 (first (dtype/shape t)) 1 false]
+    (and (nil? start) (nil? stop) step (neg? step)) ;; x[::-2]
+    [(dec (first (dtype/shape t))) 0 (Math/abs step) true]))
+
+
+(comment
+  
+  (start-stop-step-reverse (dtt/->tensor (range 20)) 10 20 nil)
+
+  ,)
+(defn
+  ^{:doc
+    "API mostly matches API of numpy, except for some corner cases,
+and an off-by-one error on some negative steps. 
+This is the most horrible thing I've ever written.  Someone please help. :(
+--JJ"}
+  array-slice
```

Comment:
I think that dtt/select will help you a lot here.  You can use ranges to do a lot of what you are trying to do and if not ranges then you can just specify the indexes verbatim per dimension.  You really do not want to be calling mget but rather working purely in index space and apply that space to the tensor all at once.

Reply:
:man_facepalming: you are absolutely right.  Even worse, I reimplemented `select` to fix this because I completely overlooked it in the API docs.

Reply:
Reimplemented using `dtt/select`, cut down on a lot of bad code.  Thanks for the tip!